cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -480,11 +480,32 @@ public void programHistory(HttpRequest request, HttpResponder responder,
                              @QueryParam("status") String status,
                              @QueryParam("start") String startTs,
                              @QueryParam("end") String endTs,
-                             @QueryParam("limit") @DefaultValue("100") final int resultLimit)
+                             @QueryParam("limit") @DefaultValue("100") final int resultLimit,
+                             @QueryParam("allVersions") Boolean allVersions)
     throws Exception {
-    programHistory(request, responder, namespaceId, appName,
-                   getLatestAppVersion(new NamespaceId(namespaceId), appName), type,
-                   programName, status, startTs, endTs, resultLimit);
+    if (allVersions != null && allVersions) {
+      // get runs of all versions
+      ProgramType programType = getProgramType(type);
+      long start = (startTs == null || startTs.isEmpty()) ? 0 : Long.parseLong(startTs);
+      long end = (endTs == null || endTs.isEmpty()) ? Long.MAX_VALUE : Long.parseLong(endTs);
+      List<RunRecord> records = new ArrayList<>();
+      Collection<ApplicationSpecification> allAppVersions = store.getAllAppVersions(
```

Comment:
We don't do pagination on get all versions? There could be a lot of versions for a single app?